src/core/lib/event_engine/thread_pool.cc

```diff
@@ -93,32 +94,41 @@ void ThreadPool::Add(absl::AnyInvocable<void()> callback) {
   }
 }
 
+bool ThreadPool::IsBusy() {
+  return state_->queue.IsBusy(state_->thread_count.threads());
+}
+
 bool ThreadPool::Queue::Add(absl::AnyInvocable<void()> callback) {
   grpc_core::MutexLock lock(&mu_);
   // Add works to the callbacks list
   callbacks_.push(std::move(callback));
   cv_.Signal();
   switch (state_) {
-    case State::kRunning:
-    case State::kShutdown:
+    case QueueState::kRunning:
+    case QueueState::kShutdown:
       return threads_waiting_ == 0;
-    case State::kForking:
+    case QueueState::kForking:
       return false;
   }
   GPR_UNREACHABLE_CODE(return false);
 }
 
-void ThreadPool::Queue::SetState(State state) {
+void ThreadPool::Queue::SetState(QueueState state) {
   grpc_core::MutexLock lock(&mu_);
-  if (state == State::kRunning) {
-    GPR_ASSERT(state_ != State::kRunning);
+  if (state == QueueState::kRunning) {
+    GPR_ASSERT(state_ != QueueState::kRunning);
   } else {
-    GPR_ASSERT(state_ == State::kRunning);
+    GPR_ASSERT(state_ == QueueState::kRunning);
   }
   state_ = state;
   cv_.SignalAll();
 }
 
+bool ThreadPool::Queue::IsBusy(int expected_idle_threads) {
```

Comment:
I don't think this is implementable once we have a work stealing implementation... Indeed I'm not convinced it is now.

The problem now:
I don't think this counts threads that are not waiting but are currently executing callbacks - for that we need to wait for thread termination.

The problem with work queues:
This gets much harder because now we can't lock the entire data structure at once, because doing so would imply having a thread pool wide mutex on the enqueue path - which we endeavor to avoid.

I suspect that the only reasonable way to get this done is to do something like the thread quiesce we do for fork, and in #30996 on the shutdown path too.